[WIP] Adding dynamic provisioner and deleter interface for glusterfs

humblec opened the pull request:
[WIP] Adding dynamic provisioner and deleter interface for GlusterFS.

Signed-off-by: hchiramm hchiramm@redhat.com

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/25026)

<!-- Reviewable:end -->

jsafrane commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -375,7 +375,8 @@ func StartControllers(s *options.CMServer, kubeClient *client.Client, kubeconfig
 
 	if provisioner != nil {
-		pvController, err := persistentvolumecontroller.NewPersistentVolumeProvisionerController(persistentvolumecontroller.NewControllerClient(clientset.NewForConfigOrDie(restclient.AddUserAgent(kubeconfig, "persistent-volume-provisioner"))), s.PVClaimBinderSyncPeriod.Duration, s.ClusterName, volumePlugins, provisioner, cloud)
+
```
nit: please remove empty line here

jsafrane commented on cmd/kube-controller-manager/app/plugins.go:
```diff
@@ -91,12 +89,14 @@ func NewVolumeProvisioner(cloud cloudprovider.Interface, config componentconfig.
-		return getProvisionablePluginFromVolumePlugins(cinder.ProbeVolumePlugins())
+	case config.EnableNetworkStorageProvisioning:
+		return getProvisionablePluginFromVolumePlugins(glusterfs.ProbeVolumePlugins())
+		//case cloud != nil && aws.ProviderName == cloud.ProviderName():
```
Do not remove cloud provisioners!

jsafrane commented on cmd/kubelet/app/options/options.go:
```diff
@@ -141,6 +141,7 @@ func NewKubeletServer() *KubeletServer {
 			KubeAPIQPS:                     5.0,
 			KubeAPIBurst:                   10,
 			ExperimentalFlannelOverlay:     experimentalFlannelOverlay,
+			StorageConfigDir:               "",
```
Why is this option needed in kubelet? It won't provision any volumes, `--storage-config=`` is relevant only to kube-controller-manager

jsafrane commented on examples/experimental/persistent-volume-provisioning/README.md:
```diff
@@ -120,6 +120,82 @@ $ kubectl get pv
+### Provision GlusterFS Volume
+Enable glusterfs dynamic provisioning by supplying `--enable-network-storage-provisioner=true --storage-config=/path/to/storage/config/directory` to `kube-controller-manager`.
+
+In the storage configuration directory, provide a gluster cluster configuration json like the following, where `endpoint` and `resturl` are mandatory.
```
Please add a note that the file must be named `gluster.json`.

jsafrane commented on hack/local-up-cluster.sh:
```diff
@@ -288,6 +291,8 @@ function start_controller_manager {
       --service-account-private-key-file="${SERVICE_ACCOUNT_KEY}" \
       --root-ca-file="${ROOT_CA_FILE}" \
       --enable-hostpath-provisioner="${ENABLE_HOSTPATH_PROVISIONER}" \
+           --enable-network-storage-provisioner="${ENABLE_NET_STORAGE_PROVISIONER}" \
```
extra spaces (or mixed spaces with tabs?)

jsafrane commented on hack/verify-flags/known-flags.txt:
```diff
@@ -101,6 +101,7 @@ e2e-verify-service-account
 enable-custom-metrics
 enable-debugging-handlers
 enable-hostpath-provisioner
+enable-glusterfs-provisioner
```
It should be `enable-network-storage-provisioner`

jsafrane commented on pkg/kubelet/kubelet.go:
```diff
@@ -213,6 +213,7 @@ func NewMainKubelet(
 	containerManager cm.ContainerManager,
 	outOfDiskTransitionFrequency time.Duration,
 	flannelExperimentalOverlay bool,
+	storageConfigDir string,
```
As pointed out above, kubelet should not need storage config directory.

jsafrane commented on pkg/kubelet/volumes.go:
```diff
@@ -104,6 +104,11 @@ func (vh *volumeHost) GetHostName() string {
 	return vh.kubelet.hostname
 }
 
+// Returns the storage config dir
```
Please do not add new functions to `VolumeHost`, it's complicated enough already - this interface is (and will be) implemented in many places.

It seems that the config dir is relevant to glusterfs and maybe to ceph. To pass the config directory to appropriate plugins you can add a parameter to `glusterfs.ProbeVolumePlugins(configDir string)`. In `kube-controller-manager` init, you will then call:

```
	allPlugins = append(allPlugins, glusterfs.ProbeVolumePlugins(config.ConfigDir)...)
```

(and you need to add `ConfigDir` to `struct VolumeConfiguration`). This requires some plumbing here and there, but it's IMO much cleaner than adding new method to `VolumeHost`.

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -260,3 +269,184 @@ func (b *glusterfsMounter) setUpAtInternal(dir string) error {
+	return d.DeleteVolume()
+}
+
+func (d *glusterfsVolumeDeleter) DeleteVolume() error {
```
What's the difference between `glusterfsVolumeDeleter.Delete` and `glusterfsVolumeDeleter.DeleteVolume`? Should it be the same method?

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -260,3 +269,184 @@ func (b *glusterfsMounter) setUpAtInternal(dir string) error {
+}
+
+func (d *glusterfsVolumeDeleter) DeleteVolume() error {
+	glog.V(1).Infof("glusterfs: delete volume :%s ", d.glusterfsMounter.path)
```
Please log normal operation with level 2.

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -260,3 +269,184 @@ func (b *glusterfsMounter) setUpAtInternal(dir string) error {
+	options volume.VolumeOptions
+}
+
+var clusterconf = new(glusterfsClusterConf)
```
No global variables please, add an item to `glusterfsPlugin`.

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -260,3 +269,184 @@ func (b *glusterfsMounter) setUpAtInternal(dir string) error {
+	file := path.Join(config, "gluster.json")
+	fp, err := os.Open(file)
+	if err != nil {
+		glog.V(1).Infof("glusterfs cluster configuration file open err %s/%s", file, err)
```
`glusterfs: error opening file %s: %s`

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -260,3 +269,184 @@ func (b *glusterfsMounter) setUpAtInternal(dir string) error {
+
+	decoder := json.NewDecoder(fp)
+	if err = decoder.Decode(clusterconf); err != nil {
+		glog.V(1).Infof("glusterfs: decode err: %s.", err)
```
`glusterfs: cannot parse file %s: %s`. It's important to show the file name, otherwise the user it lost.

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -260,3 +269,184 @@ func (b *glusterfsMounter) setUpAtInternal(dir string) error {
+func (p *glusterfsVolumeProvisioner) CreateVolume() (r *api.GlusterfsVolumeSource, size int, err error) {
+	volSizeBytes := p.options.Capacity.Value()
+	const gb = 1024 * 1024 * 1024
+	sz := int((volSizeBytes + gb - 1) / gb)
```
Please reuse `volume.RoundUpSize`

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -260,3 +269,184 @@ func (b *glusterfsMounter) setUpAtInternal(dir string) error {
+	const gb = 1024 * 1024 * 1024
+	sz := int((volSizeBytes + gb - 1) / gb)
+	glog.V(1).Infof("glusterfs: create volume of size:%s ", volSizeBytes)
+	if *(p.glusterfsClusterConf) == (glusterfsClusterConf{}) {
```
This is really ugly, check concrete `glusterfsClusterConf` items instead.

jsafrane commented on cmd/kube-controller-manager/app/plugins.go:
```diff
@@ -91,12 +89,14 @@ func NewVolumeProvisioner(cloud cloudprovider.Interface, config componentconfig.
-		return getProvisionablePluginFromVolumePlugins(gce_pd.ProbeVolumePlugins())
-	case cloud != nil && openstack.ProviderName == cloud.ProviderName():
-		return getProvisionablePluginFromVolumePlugins(cinder.ProbeVolumePlugins())
+	case config.EnableNetworkStorageProvisioning:
```
Problem with this approach is that `--enable-network-storage-provisioner` is now just for glusterfs. How are you going to decide if gluster *or* ceph should be initialized here? Current provisioner logic allows only one provisioner, ceph or gluster (or the cloud one, say AWS EBS).

humblec replied:
@jsafrane yeah, some logic has to be developed here first to decide whether its gluster /ceph or some other network storage provisioner .  Second logic has to be there to enable one cloud provisioner and one network storage provisioner at the same time. Is there any work going on in this direction? if that is the case I may need to wait for it. If there is no plan as such, Is it fine to continue on this and getting this done in this manner ? please let me know your thoughts on this, so that I can proceed accordingly.

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -260,3 +269,184 @@ func (b *glusterfsMounter) setUpAtInternal(dir string) error {
+	}
+	cli := gcli.NewClient(p.GlusterRestUrl, p.GlusterRestUser, p.GlusterRestUserKey)
+	if cli == nil {
+		glog.V(1).Infof("glusterfs: failed to create gluster rest client")
```
Is it possible to log additional details about what went wrong?

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -260,3 +269,184 @@ func (b *glusterfsMounter) setUpAtInternal(dir string) error {
+	volumeReq := &gapp.VolumeCreateRequest{Size: sz}
+	volume, err := cli.VolumeCreate(volumeReq)
+	if err != nil {
+		glog.V(1).Infof("glusterfs: error [%s] when creating the volume", err)
```
`glusterfs: error creating volume: %s`

jsafrane commented:
@humblec, thanks for your provisioner. It has some rough edges, see individual line comments, but it seems you got the provisioning logic right. However, volume plugins are undergoing significant refactoring and cleanup for attach/detach controller, see #21709 for GCE - all volume plugins will be refactored in similar way) 

Also, we need e2e test for glusterfs provisioning. We're not gluster experts here and we need an automated way how to check the provisioner/deleter is working. The test should spawn a glusterfs server in a container similar (if not identical) to gluster test in test/e2e/volumes.go and provision/use/delete a volume - add a test to test/e2e/volume_provisioning.go and disable it by default, (`--enable-network-storage-provisioner` won't be enabled by default).

humblec commented:
@jsafrane Thanks a lot for your detailed review! much appreciated. yeah, some of the hacks (wrt network provisioner ) were applied to try the logic. I will revisit all the comments  and will also work on the e2e testing part.

humblec commented:
@jsafrane I have addressed the comments  provided in your last review. Can you please validate the new patch ? e2e test part is pending though, Is it fine to send a seperate PR for e2e testing, once this is in ?

humblec committed:
[WIP ] GlusterFS dynamic provisioner and deleter interface

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
@jsafrane regarding identifying whether its gluster/ceph or some other network storage provisioner, I am thinking to introduce one more flag ( like coud provider) in struct volumeConfiguration called "net-provider" . Any thoughts?

jsafrane commented:
Sorry, been on PTO and dynamic provisioning in #29006 is not finished yet. Many things happened in the last few months and dynamic provisioning is quite different today. And it's still not stable, the PR is still on review!

In all cases, Kubernetes 1.4 should allow multiple provisioners running in parallel, so both Gluster and Ceph should be working simultaneously.  

Please do not introduce a new flag `--enable-network-storage-provisioner` to enable/disable them, they should be both enabled.

Regarding their configuration, both Gluster and Ceph can be configured in a StorageClass instances, there is `parameters` field that configure the provisioner for corresponding StoragClass. So far we use it to select volume type (e.g. SSD / HDD on GCE) and availability zone. There is an example of a StorageClass somewhere in the aforementioned PR.  It contains a free-form map[string]string with configuration for the provisioner, it could contain Ceph/Gluster address, pool name, username and such. Provisioner gets this map in `NewProvisioner` call in `VolumeOptions.Parameters` [when the PR is merged]. For password or other credentials... you can either store it the StorageClass in plaintext or point to a Secret instance. This way, Kubernetes could provision volumes using two (or even more) Ceph instances  at the same time, each would have with its own StorageClass instance with parameters specific for the particular instance.

So, please watch PR#29006 and update your provisioners accordingly. I apologize again that it takes so much time.

humblec commented:
@jsafrane first of all nice work on DP v2 PR !! yeah, looks like the input for the provisioner need to be fetched from storage class from here on.  I still have to go through the PR and understand the new bits.  I will adjust glusterfs provisioner accordingly, also keep an eye on the  mentioned PR.

humblec commented:
@jsafrane @rootfs I have rebased the PR based on DPv2 or Storage class based  claims and its available  https://github.com/humblec/kubernetes/commit/caa96c1b01655ce89b1f17dbfb07bacde69cc68b .

jsafrane commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -400,6 +400,8 @@ func StartControllers(s *options.CMServer, kubeClient *client.Client, kubeconfig
 		s.ClusterName,
 		nil, nil, nil,
 		s.VolumeConfiguration.EnableDynamicProvisioning,
+		s.VolumeConfiguration.EnableNetworkStorageProvisioning,
```
As I already wrote, **all** configuration you need should be in `StorageClass.Parameters`. Please do not introduce any new arguments.

jsafrane commented on cmd/kube-controller-manager/app/options/options.go:
```diff
@@ -75,8 +75,10 @@ func NewCMServer() *CMServer {
+				EnableHostPathProvisioning:       false,
+				EnableDynamicProvisioning:        true,
+				EnableNetworkStorageProvisioning: true,
+				StorageConfigDir:                 "/tmp",
```
dtto, no Gluster/Ceph configuration outside of StorageClass.Parameters. It's a free-form `map[string]string`, you can squeeze anything you want there.

jsafrane commented on cmd/kube-controller-manager/app/options/options.go:
```diff
@@ -128,6 +130,7 @@ func (s *CMServer) AddFlags(fs *pflag.FlagSet) {
 	fs.Int32Var(&s.VolumeConfiguration.PersistentVolumeRecyclerConfiguration.MinimumTimeoutHostPath, "pv-recycler-minimum-timeout-hostpath", s.VolumeConfiguration.PersistentVolumeRecyclerConfiguration.MinimumTimeoutHostPath, "The minimum ActiveDeadlineSeconds to use for a HostPath Recycler pod.  This is for development and testing only and will not work in a multi-node cluster.")
 	fs.Int32Var(&s.VolumeConfiguration.PersistentVolumeRecyclerConfiguration.IncrementTimeoutHostPath, "pv-recycler-timeout-increment-hostpath", s.VolumeConfiguration.PersistentVolumeRecyclerConfiguration.IncrementTimeoutHostPath, "the increment of time added per Gi to ActiveDeadlineSeconds for a HostPath scrubber pod.  This is for development and testing only and will not work in a multi-node clust ...
 	fs.BoolVar(&s.VolumeConfiguration.EnableHostPathProvisioning, "enable-hostpath-provisioner", s.VolumeConfiguration.EnableHostPathProvisioning, "Enable HostPath PV provisioning when running without a cloud provider. This allows testing and development of provisioning features.  HostPath provisioning is not supported in any way, won't work in a multi-node cluster, and should not be used for anythi ...
+	fs.BoolVar(&s.VolumeConfiguration.EnableNetworkStorageProvisioning, "enable-network-storage-provisioner", s.VolumeConfiguration.EnableNetworkStorageProvisioning, "Enable Network Storage PV provisioning.")
```
both Gluster and Ceph provisioners should be enabled by default and without any configuration. Admin then enables Ceph/Gluster by creating proper StorageClass referring the right provisioner.

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -145,6 +157,29 @@ func (plugin *glusterfsPlugin) execCommand(command string, args []string) ([]byt
 	return cmd.CombinedOutput()
 }
 
+func (plugin *glusterfsPlugin) readClusterConf() (*glusterfsClusterConf, error) {
```
you should parse `glusterfsClusterConf` out of `StorageClass.Parameters`, not from a file.

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -302,3 +337,170 @@ func getVolumeSource(
+	return pv, nil
+}
+
+func (c *glusterfsVolumeProvisioner) NewPersistentVolumeTemplate() (*api.PersistentVolume, error) {
```
Make this function private and do not return error if you don't use it.

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -302,3 +337,170 @@ func getVolumeSource(
+	options volume.VolumeOptions
+}
+
+func (plugin *glusterfsPlugin) NewDeleter(spec *volume.Spec) (volume.Deleter, error) {
```
umm, if all the configuration  parameters are in StorageClass, then `Deleter` does not have access to it. I'll add `Parameters` argument here when DPv2 is merged.

jsafrane replied:
filled #30864

humblec commented:
Closing this PR and new one based on storage class is available @ PR#30888
